### ISSUE TITLE
WER increases with increasing beam size, and with including LM

### ISSUE THREAD

**Issue by shrutijpalaskar** [espnet/espnet#135]
Hello,

I am running experiments with the WSJ corpus using PyTorch. I only use the seq2seq part of the code with --mtlalpha set to 0. During decoding, even without a language model, the CER, WER with greedy decoding is best and increasing the beam size only leads to increased errors.

Upon LM training on the WSJ corpus, the best model perplexity reached is: 
``` 
2018-04-04 21:44:27,006 (lm_pytorch:248) INFO: iteration: 79400
2018-04-04 21:44:27,006 (lm_pytorch:249) INFO: training perplexity: 2.81606144442938
2018-04-04 21:45:16,637 (lm_pytorch:255) INFO: epoch: 24
2018-04-04 21:45:16,638 (lm_pytorch:256) INFO: validation perplexity: 4.354437210237004 
```

And decoding with including such a LM, worsens CER, WER results further.

I believe this might be due to some issue in the decoder or the beam search? Please let me know if you need any further details.

**Comment by sw005320** [comment]
Thanks for your report.
Yes, it must be a bug.
I'll check it soon.

**Comment by ShigekiKarita** [comment]
@shrutijpalaskar  I will try to reproduce this. Could you share us your decoding parameters (e.g., beam size, lmweight) and results (e.g., CER, WER)?

**Comment by shrutijpalaskar** [comment]
@ShigekiKarita thank you for looking into this. The parameter values are as follows:

backend = pytorch
mtl alpha = 0
etype = blstmp
bs = 48
ls = unigram 0.05
beam size = 20 
penalty = 0.0
max min ratio = 0.0, 0.0
ctc weight = 0
rnn_lm weights = 0.0, 0.3 and 1.0

And the results where we see that adding LM worsens performance are as follows:
** dev93 **
1. decode_test_dev93_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0_rnnlm0.0 : CER=11.6, WER=25.3
2. decode_test_dev93_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0_rnnlm0.3 : CER=22.3, WER=31.6
3. decode_test_dev93_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0_rnnlm1.0 : CER=56.2, WER=61.6

**eval92**
1. decode_test_eval92_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0_rnnlm0.0 : CER=6.9, WER=18.3
2. decode_test_eval92_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0_rnnlm0.3 : CER=23.8, WER=30.1
3. decode_test_eval92_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0_rnnlm1.0 : CER=57.4, WER=61

**Comment by ShigekiKarita** [comment]
@shrutijpalaskar thanks

https://github.com/espnet/espnet/blob/4d6b1ecb430aa1701ffc6f4ad9f6d2c9b26428d4/src/nets/e2e_asr_attctc_th.py#L1725-L1726

as you can see here, lm_weight=1.0 ignores encoder-decoder score. i.e. no speech inputs. so larger LM weight does not always improve ASR because it is just a kind of hyper parameter.

btw, could you share us the results with larger beam-sizes? I hope the larger beam-size should be better unlike the LM weight.

**Comment by sw005320** [comment]
> @shrutijpalaskar thanks
> 
> espnet/src/nets/e2e_asr_attctc_th.py
> 
> Lines 1725 to 1726 in 4d6b1ec
> 
>  final_z = (1 - recog_args.lm_weight) * F.log_softmax(self.output(z_list[-1])) \ 
>      + recog_args.lm_weight * F.log_softmax(z_rnnlm) 
> as you can see here, lm_weight=1.0 ignores encoder-decoder score. i.e. no speech inputs. so larger LM weight does not always improve ASR because it is just a kind of hyper parameter.

This is the case without beam search (greedy search). It seems that I did not use the same way for `recog_args.lm_weight` between beam search and greedy search. I'll fix it.
Also, I'll check @m-wiesner's report for the bug in the greedy search.

**Comment by sw005320** [comment]
@m-wiesner, At least the problem of the beam size 1 would be fixed at #149.
@shrutijpalaskar, I'll now move on your problem. Sorry for my late response.

**Comment by shrutijpalaskar** [comment]
Thanks a lot for looking into it @sw005320 

I will get the results for varying beam sizes this weekend.

**Comment by shrutijpalaskar** [comment]
Hello, @sw005320 @ShigekiKarita @m-wiesner,

Here are the results for varying beam size with WSJ. The error decreases till beam=5 and then increases with increasing beam size. The vertical axis shows character or word error rate and the horizontal axis shows beam size. My experiments parameters are: 

backend = pytorch
mtl alpha = 0
etype = blstmp
bs = 48
ls = unigram 0.05
penalty = 0.0
max min ratio = 0.0, 0.0
ctc weight = 0
rnn_lm weights = 0.0

![chart](https://user-images.githubusercontent.com/11374673/40258824-eb12e8c8-5ac0-11e8-92fc-6133735467e9.png)

**Comment by sw005320** [comment]
Thanks for the report! This is quite valuable analysis.
One thing I want to mention is that the wide beam without any decoding parameter tuning tends to provide crazy  results (this was solved by using a joint CTC/attention decoding or setting max min ratios). I need to take a look at your results in more details. Could you provide either WER/CER with substitution, deletion, and insertion rates at the beam width of 1, 5, and 20? Especially if we have some unbalance deletion and insertion numbers, which indicate the problem I mentioned above.

**Comment by shrutijpalaskar** [comment]
hello @sw005320 here are the detailed results:

It has an increase in number of insertions as you said
CER Dev93:
```
decode_test_dev93_beam1_p0_len0.0-0.0_nj8/result.txt:|         Sum/Avg        |          503                   48634         |         92.8                    4.0                    3.2                    2.4                    9.6                   93.0         |
decode_test_dev93_beam5_p0_len0.0-0.0_nj8/result.txt:|         Sum/Avg        |          503                   48634         |         93.0                    3.6                    3.4                    1.9                    8.9                   91.3         |
decode_test_dev93_beam20_p0_len0.0-0.0_nj8/result.txt:|         Sum/Avg        |          503                   48634         |         90.1                    3.4                     6.5                    1.7                   11.6                   91.1         |
```

WER Dev93:
```
decode_test_dev93_beam1_p0_len0.0-0.0_nj8/result.wrd.txt:|         Sum/Avg         |          503                   8234          |         78.0                   19.4                     2.6                    3.3                    25.3                   92.0          |
decode_test_dev93_beam5_p0_len0.0-0.0_nj8/result.wrd.txt:|         Sum/Avg         |          503                   8234          |         79.4                   17.9                     2.7                    2.7                    23.3                   90.3          |
decode_test_dev93_beam20_p0_len0.0-0.0_nj8/result.wrd.txt:|         Sum/Avg         |          503                   8234          |         77.2                    17.2                    5.6                     2.4                    25.3                   90.3          |
```

CER Eval92:
```
decode_test_eval92_beam1_p0_len0.0-0.0_nj8/result.txt:|         Sum/Avg        |          333                   33341         |         95.2                    2.8                     1.9                    1.9                    6.7                   91.0         |
decode_test_eval92_beam5_p0_len0.0-0.0_nj8/result.txt:|         Sum/Avg        |          333                   33341         |         95.3                    2.6                     2.1                    1.7                    6.4                   89.2         |
decode_test_eval92_beam20_p0_len0.0-0.0_nj8/result.txt:|         Sum/Avg        |          333                   33341          |         94.7                    2.5                    2.8                    1.6                     6.9                   89.2         |
```

WER Eval92:
```
decode_test_eval92_beam1_p0_len0.0-0.0_nj8/result.wrd.txt:|         Sum/Avg         |          333                   5643          |         82.9                    15.8                    1.4                     2.5                    19.6                   89.5          |
decode_test_eval92_beam5_p0_len0.0-0.0_nj8/result.wrd.txt:|         Sum/Avg         |          333                   5643          |         84.5                    14.0                    1.4                     2.5                    18.0                   87.1          |
decode_test_eval92_beam20_p0_len0.0-0.0_nj8/result.wrd.txt:|         Sum/Avg         |          333                    5643         |         84.1                    13.7                     2.1                     2.4                   18.3                    87.1          |
```

**Comment by atiklt1972** [comment]
Hello,
I am running an experiment using my native language dataset . I used the espnet speech processing toolkit by configuring the parameters only. Phoneme,character and word level language units with phoneme and character acoustic modeling units are used in my experiment.my expectation; the WER  could minimum when I used phoneme and character LM units rather than word LM.The result found minimum WER in word LM but CERs are as my expectation. please anyone tell me the problem.

**Comment by atiklt1972** [comment]
waiting@sw005320

**Comment by sw005320** [comment]
A word-based model would have OOV issues, and you would not always get better results from character-based LM. Our word-based LM implemented in espnet (you can find it in the WSJ recipe) is to solve this problem by training ASR models with a character unit but using the word-based LM during inference.

**Comment by atiklt1972** [comment]
Thak you @sw005320. I will check it and I will reach out  you .

**Comment by atiklt1972** [comment]
In character RNNLM
          CTC | 65.7
         Attention | 69.6
         Hybrid | 57.1
In word RNNLM
        CTC | 58.5
       Attention | 58.5
      Hybrid | 33.3
@sw005320 this was the result for your information.I have two basic quesitions.1)why WER is more inCharacter LM than wordRNNLM? 2)why error rate is more in location based attention as compared LAS architecture?
my expectation was OOV of words in word based LM but the result is the opposite.

**Comment by atiklt1972** [comment]
Hi@sw005320 ,
Most of the utterances/sentences of my dataset are too long but the speech recognition toolkit (spent) expects relatively shorter utterances. I think errors are more due to long utterances and I observed that recognition error differences between short and long ones. I want to use the speech segmentation techniques in order to make short utterances. what tools are advised for automatic speech segmentation as part of data preprocessing? please add more advice.

**Comment by sw005320** [comment]
This is one of the missing parts of espnet but in general ASR toolkits are operated with internal/external speech activity detection (SAD) tools.
You can use Kaldi SAD
Search "Kaldi" and "SAD"

**Comment by atiklt1972** [comment]
Ok, Thanks for your immediate response.
